guides/create_an_app/4-query.md

```diff
@@ -1,5 +1,88 @@
+      {:ok, socket}
+    else
+      {:error, %{reason: "unauthorized"}}
+    end
+  end
+```
+
+### What does send do?
+`send` sends a message `:after_join`, with it's PID generated by the function `self()`. We want to be able to run that after join task asynchronously. Lets look at how we should receive the `:after_join` message.
+
+## Receiving a message in a callback
+We now know that a message identified by the atom `:after_join` is being sent. Phoenix provides us with a function called `handle_info` to be able to receive that message and do something with it. Insert the following code after the `join` function.
+
+```elixir
+  def handle_info(:after_join, socket) do
+    messages =
+      WorkshopsApp.Conversations.list_messages()
+      |> Enum.map(fn(m) -> %{message: m.message, name: m.name} end)
+
+    push socket, "messages_history", %{messages: messages}
+    {:noreply, socket}
+  end
+```
+
+The only missing thing is `WorkshopsApp.Conversations.list_messages()` (we should get compilation error when function is called but not definied). To have everything working as expected, we need to add function, which will list all messages stored in our database. Let's open `WorkshopsApp.Conversations` module and put there:
+
+```elixir
+def list_messages do
+  Repo.all(Message)
+end
+
+```
+
+**Coach:** WDYT about saying couple words about compilator?
+
+## Retrieving all messages from our database
+Phew, that looked like a lot is going on in the `handle_info` function. Let's breakdown the code to understand what is going on.
+
+Thanks to `list_messages()` function we can retrieve all the messages from our database.
+
+We then, pipe the retrieved messages into a `map()` function provided by the `Enum` module. `map` applies this anonymous function onto every element in messages.
+
+**Coach:** Please explain how Enum.map works if it's unsure :) also it's good time to play with other functions from Enum module.
+
+**Coach:** Please explain `&` operator for function shorthands possibility.
+
+## Passing the messages back to the browser
+Finally, we want to push our list of messages to the socket via the push function.
+
+We are using `push` instead of `broadcast`, because we only want to display it to the current user that has just joined the channel.
+
+## Render the messages in the browser
+Add the below code to the `assets/js/socket.js` file.
+
+```javascript
+    channel.on('messages_history', messages => {
+      let messages_list = messages["messages"];
+
+      messages_list.forEach( function(msg) {
+        list.append(`<b>${msg["name"] || 'Anonymous'}:</b> ${msg["message"]}<br>`);
+        list.prop({scrollTop: list.prop("scrollHeight")});
+      });
+    });
+```
+
+This is the final step, really. We need display the messages in our browser by iterating through our list You should have the history of messages displayed to you when you join the channel. Try opening multiple tabs connecting to http://localhost/4000
```

Comment:
This is the final step, really. We need to display the messages in our browser by iterating through our list. You should have the history of messages displayed the moment you join the channel. Try opening multiple tabs connecting to http://localhost/4000.